### PR TITLE
#61: Create configuration validator

### LINKED ISSUE #61: Create configuration validator
This configuration validator should ensure that the visualization parameters are correct, cf. https://github.com/DARMA-tasking/LB-analysis-framework/issues/469

### PR DESCRIPTION
**Pull request by maxime-bfsquall**
Fixes #61

### CODE REVIEW

File: bindings/python/config_validator.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
+/*
+//@HEADER
+// *****************************************************************************
+//
+//                               config_validator.h
+//             DARMA/vt-tv => Virtual Transport -- Task Visualizer
+//
+// Copyright 2019 National Technology & Engineering Solutions of Sandia, LLC
```
Comment (ppebay) [inline bindings/python/config_validator.h:8]: Update date here

File: bindings/python/tv.cc [hunk 2]
```diff
@@ -17,6 +18,16 @@ void tvFromJson(const std::vector<std::string>& input_json_per_rank_list, const
     // Load the configuration from serialized YAML
     YAML::Node viz_config = YAML::Load(input_yaml_params_str);
 
+    // Config Validator
+    ConfigValidator config_validator(viz_config);
+
+    // Check configuration
+    bool is_config_valid = config_validator.isValid();
+
+    // Throw error if configuration is not valid
```
Comment (ppebay) [inline bindings/python/tv.cc:27]: `is invalid`

File: tests/test_bindings.py [hunk 3]
```diff
@@ -1,33 +1,47 @@
...
-    params = yaml.safe_load(stream)
-  except yaml.YAMLError as exc:
-    print(exc)
+    try:
+        params = yaml.safe_load(stream)
+    except yaml.YAMLError as exc:
+        print(exc)
+
+# Check main key is "visualization"
+if 'visualization' not in params:
```
Comment (ppebay) [inline tests/test_bindings.py:21]: Code style for Python in LBAF ecosystem is to conform to C-style simple quotes for single characters, vs. double quotes for string literals. I would suggest respecting the same style here for consistency.

File: tests/test_bindings.py [hunk 4]
```diff
@@ -1,33 +1,47 @@
...
+if 'output_visualization_dir' in params["visualization"]:
+    params["visualization"]["output_visualization_dir"] = os.path.abspath(
+      params["visualization"]["output_visualization_dir"]
+    )
 
 params_serialized = yaml.dump(params["visualization"])
 
-n_ranks = params["visualization"]["x_ranks"] * params["visualization"]["y_ranks"] * params["visualization"]["z_ranks"]
-rank_data = []
+n_ranks = params["visualization"]["x_ranks"]
```
Comment (ppebay) [inline tests/test_bindings.py:33]: What is the reason for separating this in 3 lines, requiring 2 additional assignments?

File: tests/test_bindings.py [hunk 5]
```diff
@@ -1,33 +1,47 @@
...
-
-  data_serialized = json.dumps(data)
+    with open(f'{source_dir}/data/lb_test_data/data.{rank}.json', 'r') as f:
+        data = json.load(f)
 
-  rank_data.append((data_serialized))
+    data_serialized = json.dumps(data)
+    rank_data.append((data_serialized))
 
+# Launch
```
Comment (ppebay) [inline tests/test_bindings.py:46]: Specify what is launched

**Review by ppebay — CHANGES_REQUESTED** [review]

File: tests/test_bindings.py (hunk 4 above)
Comment (maxime-bfsquall) [reply tests/test_bindings.py:33]: Pylint warning: Line too long

File: bindings/python/config_validator.h [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+*/
+// A2DD.h
+#ifndef vt_tv_config_validator_h
+#define vt_tv_config_validator_h
+
+#include <yaml-cpp/yaml.h>
+
+namespace vt::tv::bindings::python {
+    /**
+     * ConfiValidator Class
```
Comment (tlamonthezie) [inline bindings/python/config_validator.h:51]: Please fix typo

File: tests/test_bindings.py [hunk 7]
```diff
@@ -1,33 +1,47 @@
...
+n_ranks *= params["visualization"]["z_ranks"]
 
+# Prepare rank data
+rank_data = []
 for rank in range(n_ranks):
-  with open(f'{source_dir}/data/lb_test_data/data.{rank}.json', 'r') as f:
-    data = json.load(f)
-
-  data_serialized = json.dumps(data)
+    with open(f'{source_dir}/data/lb_test_data/data.{rank}.json', 'r') as f:
```
Comment (tlamonthezie) [inline tests/test_bindings.py:40]: I think you might have a pylint warning here. (specify encoding to be utf-8 in the `open` arguments).

File: tests/test_bindings.py [hunk 8]
```diff
@@ -1,33 +1,47 @@
+'''TESTS BINDING'''
```
Comment (tlamonthezie) [inline tests/test_bindings.py:1]: Please use double quotes for python doc strings

File: bindings/python/config_validator.h [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+     */
+    class ConfigValidator
+    {
+        public:
+            std::array<std::string, 2> requiredParameters = {"output_visualization_dir", "output_visualization_file_stem"};
```
Comment (tlamonthezie) [inline bindings/python/config_validator.h:56]: In other C++ classes in `vt-tv`we use snake case attribute names. Please rename to `required_parameters` to be consistent.

File: bindings/python/config_validator.h [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+            YAML::Node config;
+            bool isValid();
+            std::string getMissingRequiredParameters();
+            ConfigValidator(YAML::Node configData) {
```
Comment (tlamonthezie) [inline bindings/python/config_validator.h:60]: For consistency I think (as e.g. in the `ObjectInfo` class)
- it is better to use the `in_` parameter prefix in constructor `ConfigValidator(YAML::Node in_config)` (and name the attribute without the `in_` prefix)
- use the parameter initialization feature as possible
- use private attribute as possible (with getter if needed)
```c++
ConfigValidator(YAML::Node in_config) 
  :config(in_config) {}
```

**Review by tlamonthezie — CHANGES_REQUESTED** [review]

**Comment by tlamonthezie** [comment]
@maxime-bfsquall Could you please check ci failures. There seems to be some python error.

**Review by ppebay — APPROVED** [review]

**Comment by ppebay** [comment]
@tlamonthezie I am fine with the current version of the PR; are you still seeing the CI failures you were describing previously
?
> @maxime-bfsquall Could you please check ci failures. There seems to be some python error.

File: bindings/python/tv.cc [hunk 11]
```diff
@@ -17,6 +18,16 @@ void tvFromJson(const std::vector<std::string>& input_json_per_rank_list, const
...
 
+    // Config Validator
+    ConfigValidator config_validator(viz_config);
+
+    // Check configuration
+    bool is_config_valid = config_validator.isValid();
+
+    // Throw error if configuration is invalid
+    if (!is_config_valid) {
+      throw std::runtime_error("The YAML configuration file is not valid: missing required paramaters: " + config_validator.getMissingRequiredParameters());
```
Comment (tlamonthezie) [inline bindings/python/tv.cc:29]: To be consistent between error messages I suggest to add "vt-tv: " prefix and update the message as the following:
`"vt-tv: Error validating the configuration file: " + config_validator.getMissingRequiredParameters());"`

**Review by tlamonthezie — CHANGES_REQUESTED** [review]

File: tests/test_bindings.py [hunk 12]
```diff
@@ -1,38 +1,45 @@
...
     try:
         params = yaml.safe_load(stream)
     except yaml.YAMLError as exc:
         print(exc)
-        exit(1)
+
+# Check main key is "visualization"
+if "visualization" not in params:
+    print("The YAML configuration file is not valid: missing required paramaters: visualization")
+    sys.exit(1)
```
[the remark this answers is not in the document]
Comment (maxime-bfsquall) [reply tests/test_bindings.py:21]: image
https://pylint.readthedocs.io/en/latest/user_guide/messages/refactor/consider-using-sys-exit.html

**Review by tlamonthezie — APPROVED** [review]

**Pull request merged**